### DISCUSSION TITLE
Immediate crash first time opening if home directory not writable

### DISCUSSION THREAD

**Discussion by yochem** [ghostty-org/ghostty#3163] [Feature Requests, Ideas] [answered]
If the (current users) home directory is not writable (i.e. `chmod -w $HOME`), Ghostty is never able to start for the first time after downloading the binary. After making the home directory writable, opening Ghostty for the first time, and making the home directory unwritable again, Ghostty keeps on working. This is probably due to some initializations it does the first time opening.

### Steps to reproduce

1. Download the MacOS binary from [the website](https://ghostty.org/download) and do the MacOS install things
2. Make your home directory unwritable (`chmod -w $HOME`)
3. Open Ghostty
4. Immediate crash
5. Make home dir writable again (`chmod +w $HOME`)
6. Open Ghostty again
7. Works!

### Crash report and version info

- Ghostty: v1.0.0 (I assume)
- MacOS v14.7.1 (more details in the crash report)

<details><summary>Full crash report</summary>
<pre>
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Process:               ghostty [6159]
Path:                  /Applications/Ghostty.app/Contents/MacOS/ghostty
Identifier:            com.mitchellh.ghostty
Version:               1.0.0 (8469)
Code Type:             X86-64 (Native)
Parent Process:        launchd [1]
User ID:               501

Date/Time:             2024-12-26 23:01:03.1917 +0100
OS Version:            macOS 14.7.1 (23H222)
Report Version:        12
Bridge OS Version:     9.1 (22P1072)
Anonymous UUID:        D2944187-D1CA-1E11-E043-7CD93E73EDB4

Time Awake Since Boot: 1500 seconds

System Integrity Protection: enabled

Crashed Thread:        5

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000

Termination Reason:    Namespace SIGNAL, Code 6 Abort trap: 6
Terminating Process:   ghostty [6159]

Application Specific Information:
abort() called

Thread 0::  Dispatch queue: com.apple.main-thread
0   ghostty                       	       0x10b5aecc4 std::__1::unordered_map<char const*, int, (anonymous namespace)::str_hash, (anonymous namespace)::str_eq, std::__1::allocator<std::__1::pair<char const* const, int>>>::operator[](char const*&&) + 20

Thread 1:
0   libsystem_pthread.dylib       	    0x7ff810f70ac0 start_wqthread + 0

Thread 2:
0   libsystem_pthread.dylib       	    0x7ff810f70ac0 start_wqthread + 0

Thread 3:
0   libsystem_pthread.dylib       	    0x7ff810f70ac0 start_wqthread + 0

Thread 4:
0   libsystem_pthread.dylib       	    0x7ff810f70ac0 start_wqthread + 0

Thread 5 Crashed:
0   libsystem_kernel.dylib        	    0x7ff810f3bd96 __pthread_kill + 10
1   libsystem_pthread.dylib       	    0x7ff810f74ebd pthread_kill + 262
2   libsystem_c.dylib             	    0x7ff810e9aa79 abort + 126
3   libsystem_malloc.dylib        	    0x7ff810d9f3a9 malloc_vreport + 857
4   libsystem_malloc.dylib        	    0x7ff810da297a malloc_report + 151

Thread 5 crashed with X86 Thread State (64-bit):
  rax: 0x0000000000000000  rbx: 0x0000000000000006  rcx: 0x000070000de797a8  rdx: 0x0000000000000000
  rdi: 0x0000000000007b03  rsi: 0x0000000000000006  rbp: 0x000070000de797d0  rsp: 0x000070000de797a8
   r8: 0x000000000000002e   r9: 0x0000000000000000  r10: 0x0000000000000000  r11: 0x0000000000000246
  r12: 0x0000000000007b03  r13: 0x0000000000000050  r14: 0x000070000de7a000  r15: 0x0000000000000016
  rip: 0x00007ff810f3bd96  rfl: 0x0000000000000246  cr2: 0x0000000000000000
  
Logical CPU:     0
Error Code:      0x02000148 
Trap Number:     133

Binary Images:
       0x10c882000 -        0x10c8c5fff org.sparkle-project.Sparkle (2.6.3) <3c83378b-3c44-3675-9d9c-ded47fe734ea> /Applications/Ghostty.app/Contents/Frameworks/Sparkle.framework/Versions/B/Sparkle
       0x10b45e000 -        0x10c4e7fff com.mitchellh.ghostty (1.0.0) <f263bce6-c04b-3a3f-9a6b-297a11b22b9a> /Applications/Ghostty.app/Contents/MacOS/ghostty
               0x0 - 0xffffffffffffffff ??? (*) <00000000-0000-0000-0000-000000000000> ???
    0x7ff810f6f000 -     0x7ff810f7aff7 libsystem_pthread.dylib (*) <79ecab15-71f1-3d6a-8a96-0623e622205f> /usr/lib/system/libsystem_pthread.dylib
    0x7ff810f34000 -     0x7ff810f6eff7 libsystem_kernel.dylib (*) <b8a8e52d-b216-3bdb-a31b-1e34e9e9500b> /usr/lib/system/libsystem_kernel.dylib
    0x7ff810e1b000 -     0x7ff810ea2ff7 libsystem_c.dylib (*) <17b641ba-925c-39a9-aa43-ab3b0bcdfe01> /usr/lib/system/libsystem_c.dylib
    0x7ff810d91000 -     0x7ff810dd0fff libsystem_malloc.dylib (*) <eff78fb7-b5a3-3efc-8c45-309184ae989f> /usr/lib/system/libsystem_malloc.dylib

External Modification Summary:
  Calls made by other processes targeting this process:
    task_for_pid: 0
    thread_create: 0
    thread_set_state: 0
  Calls made by this process:
    task_for_pid: 0
    thread_create: 0
    thread_set_state: 0
  Calls made by all processes on this machine:
    task_for_pid: 0
    thread_create: 0
    thread_set_state: 0

VM Region Summary:
ReadOnly portion of Libraries: Total=750.8M resident=0K(0%) swapped_out_or_unallocated=750.8M(100%)
Writable regions: Total=576.0M written=0K(0%) resident=0K(0%) swapped_out=0K(0%) unallocated=576.0M(100%)

                                VIRTUAL   REGION 
REGION TYPE                        SIZE    COUNT (non-coalesced) 
===========                     =======  ======= 
Activity Tracing                   256K        1 
ColorSync                          156K       22 
Kernel Alloc Once                    8K        1 
MALLOC                           549.3M       23 
MALLOC guard page                   32K        8 
STACK GUARD                       56.0M        6 
Stack                             26.0M        6 
VM_ALLOCATE                        268K        4 
__CTF                               824        1 
__DATA                            28.0M      443 
__DATA_CONST                      28.4M      273 
__DATA_DIRTY                      1120K      137 
__FONT_DATA                        2352        1 
__LINKEDIT                       183.4M        4 
__OBJC_RO                         71.9M        1 
__OBJC_RW                         2201K        2 
__TEXT                           567.4M      463 
mapped file                       53.4M       10 
shared memory                      784K       17 
===========                     =======  ======= 
TOTAL                              1.5G     1423 

-----------
Full Report
-----------

{"app_name":"ghostty","timestamp":"2024-12-26 23:01:03.00 +0100","app_version":"1.0.0","slice_uuid":"f263bce6-c04b-3a3f-9a6b-297a11b22b9a","build_version":"8469","platform":1,"bundleID":"com.mitchellh.ghostty","share_with_app_devs":0,"is_first_party":0,"bug_type":"309","os_version":"macOS 14.7.1 (23H222)","roots_installed":0,"name":"ghostty","incident_id":"67019F76-94B6-4DED-BDA7-5FEA3069A164"}
{
  "uptime" : 1500,
  "procRole" : "Foreground",
  "version" : 2,
  "userID" : 501,
  "deployVersion" : 210,
  "modelCode" : "MacBookAir8,1",
  "coalitionID" : 1837,
  "osVersion" : {
    "train" : "macOS 14.7.1",
    "build" : "23H222",
    "releaseType" : "User"
  },
  "captureTime" : "2024-12-26 23:01:03.1917 +0100",
  "codeSigningMonitor" : 0,
  "incident" : "67019F76-94B6-4DED-BDA7-5FEA3069A164",
  "pid" : 6159,
  "cpuType" : "X86-64",
  "roots_installed" : 0,
  "bug_type" : "309",
  "procLaunch" : "2024-12-26 23:01:02.9599 +0100",
  "procStartAbsTime" : 1555404294075,
  "procExitAbsTime" : 1555635317238,
  "procName" : "ghostty",
  "procPath" : "\/Applications\/Ghostty.app\/Contents\/MacOS\/ghostty",
  "bundleInfo" : {"CFBundleShortVersionString":"1.0.0","CFBundleVersion":"8469","CFBundleIdentifier":"com.mitchellh.ghostty"},
  "storeInfo" : {"deviceIdentifierForVendor":"6DCCE2BE-B77E-5272-8B43-895337D353B0","thirdParty":true},
  "parentProc" : "launchd",
  "parentPid" : 1,
  "coalitionName" : "com.mitchellh.ghostty",
  "crashReporterKey" : "D2944187-D1CA-1E11-E043-7CD93E73EDB4",
  "codeSigningID" : "com.mitchellh.ghostty",
  "codeSigningTeamID" : "24VZTF6M5V",
  "codeSigningFlags" : 570503953,
  "codeSigningValidationCategory" : 6,
  "codeSigningTrustLevel" : 4294967295,
  "bridgeVersion" : {"build":"22P1072","train":"9.1"},
  "sip" : "enabled",
  "exception" : {"codes":"0x0000000000000000, 0x0000000000000000","rawCodes":[0,0],"type":"EXC_CRASH","signal":"SIGABRT"},
  "termination" : {"flags":0,"code":6,"namespace":"SIGNAL","indicator":"Abort trap: 6","byProc":"ghostty","byPid":6159},
  "asi" : {"libsystem_c.dylib":["abort() called"]},
  "extMods" : {"caller":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"system":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"targeted":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"warnings":0},
  "faultingThread" : 5,
  "threads" : [{"id":45714,"threadState":{"r13":{"value":37},"rax":{"value":4490898681},"rflags":{"value":66054},"cpu":{"value":3},"r14":{"value":140701864691872},"rsi":{"value":140701864691776},"r8":{"value":3},"cr2":{"value":4490898681},"rdx":{"value":382},"r10":{"value":32},"r9":{"value":379},"r15":{"value":140704512129392,"symbolLocation":192,"symbol":"current_categories"},"rbx":{"value":105553146045824},"trap":{"value":14,"description":"(no mapping for user data read)"},"err":{"value":4},"r11":{"value":105553135644512},"rip":{"value":4485475524},"rbp":{"value":4489415552,"symbolLocation":0,"symbol":"io.GenericWriter(*io.counting_writer.CountingWriter(io.GenericWriter(void,error{},(function 'dummyWrite'))),error{},(function 'write')).typeErasedWriteFn"},"rsp":{"value":140701864691696},"r12":{"value":4490898681},"rcx":{"value":58185824},"flavor":"x86_THREAD_STATE","rdi":{"value":105553146045824}},"queue":"com.apple.main-thread","frames":[{"imageOffset":1379524,"symbol":"std::__1::unordered_map<char const*, int, (anonymous namespace)::str_hash, (anonymous namespace)::str_eq, std::__1::allocator<std::__1::pair<char const* const, int>>>::operator[](char const*&&)","symbolLocation":20,"imageIndex":1}]},{"id":45723,"frames":[{"imageOffset":6848,"symbol":"start_wqthread","symbolLocation":0,"imageIndex":3}],"threadState":{"r13":{"value":0},"rax":{"value":33554800},"rflags":{"value":512},"cpu":{"value":0},"r14":{"value":1},"rsi":{"value":4099},"r8":{"value":409602},"cr2":{"value":0},"rdx":{"value":123145516670976},"r10":{"value":0},"r9":{"value":18446744073709551615},"r15":{"value":123145517194104},"rbx":{"value":123145517195264},"trap":{"value":133},"err":{"value":33554800},"r11":{"value":582},"rip":{"value":140703413242560},"rbp":{"value":0},"rsp":{"value":123145517195264},"r12":{"value":5193730},"rcx":{"value":0},"flavor":"x86_THREAD_STATE","rdi":{"value":123145517195264}}},{"id":45724,"frames":[{"imageOffset":6848,"symbol":"start_wqthread","symbolLocation":0,"imageIndex":3}],"threadState":{"r13":{"value":0},"rax":{"value":33554800},"rflags":{"value":512},"cpu":{"value":0},"r14":{"value":1},"rsi":{"value":8963},"r8":{"value":5193730},"cr2":{"value":0},"rdx":{"value":123145517207552},"r10":{"value":0},"r9":{"value":2},"r15":{"value":123145517730680},"rbx":{"value":123145517731840},"trap":{"value":133},"err":{"value":33554800},"r11":{"value":582},"rip":{"value":140703413242560},"rbp":{"value":0},"rsp":{"value":123145517730320},"r12":{"value":5193730},"rcx":{"value":123145517730688},"flavor":"x86_THREAD_STATE","rdi":{"value":123145517731840}}},{"id":45725,"frames":[{"imageOffset":6848,"symbol":"start_wqthread","symbolLocation":0,"imageIndex":3}],"threadState":{"r13":{"value":0},"rax":{"value":33554800},"rflags":{"value":512},"cpu":{"value":0},"r14":{"value":1},"rsi":{"value":10243},"r8":{"value":409604},"cr2":{"value":0},"rdx":{"value":123145517744128},"r10":{"value":0},"r9":{"value":18446744073709551615},"r15":{"value":123145518267256},"rbx":{"value":123145518268416},"trap":{"value":133},"err":{"value":33554800},"r11":{"value":582},"rip":{"value":140703413242560},"rbp":{"value":0},"rsp":{"value":123145518268416},"r12":{"value":5193732},"rcx":{"value":0},"flavor":"x86_THREAD_STATE","rdi":{"value":123145518268416}}},{"id":45728,"frames":[{"imageOffset":6848,"symbol":"start_wqthread","symbolLocation":0,"imageIndex":3}],"threadState":{"r13":{"value":0},"rax":{"value":33554800},"rflags":{"value":512},"cpu":{"value":0},"r14":{"value":0},"rsi":{"value":18179},"r8":{"value":409604},"cr2":{"value":0},"rdx":{"value":123145518280704},"r10":{"value":0},"r9":{"value":18446744073709551615},"r15":{"value":0},"rbx":{"value":123145518804992},"trap":{"value":133},"err":{"value":33554800},"r11":{"value":582},"rip":{"value":140703413242560},"rbp":{"value":0},"rsp":{"value":123145518804992},"r12":{"value":0},"rcx":{"value":0},"flavor":"x86_THREAD_STATE","rdi":{"value":123145518804992}}},{"triggered":true,"id":45734,"threadState":{"r13":{"value":80},"rax":{"value":0},"rflags":{"value":582},"cpu":{"value":0},"r14":{"value":123145535594496},"rsi":{"value":6},"r8":{"value":46},"cr2":{"value":0},"rdx":{"value":0},"r10":{"value":0},"r9":{"value":0},"r15":{"value":22},"rbx":{"value":6},"trap":{"value":133},"err":{"value":33554760},"r11":{"value":582},"rip":{"value":140703413026198,"matchesCrashFrame":1},"rbp":{"value":123145535592400},"rsp":{"value":123145535592360},"r12":{"value":31491},"rcx":{"value":123145535592360},"flavor":"x86_THREAD_STATE","rdi":{"value":31491}},"frames":[{"imageOffset":32150,"symbol":"__pthread_kill","symbolLocation":10,"imageIndex":4},{"imageOffset":24253,"symbol":"pthread_kill","symbolLocation":262,"imageIndex":3},{"imageOffset":522873,"symbol":"abort","symbolLocation":126,"imageIndex":5},{"imageOffset":58281,"symbol":"malloc_vreport","symbolLocation":857,"imageIndex":6},{"imageOffset":72058,"symbol":"malloc_report","symbolLocation":151,"imageIndex":6}]}],
  "usedImages" : [
  {
    "source" : "P",
    "arch" : "x86_64",
    "base" : 4505214976,
    "CFBundleShortVersionString" : "2.6.3",
    "CFBundleIdentifier" : "org.sparkle-project.Sparkle",
    "size" : 278528,
    "uuid" : "3c83378b-3c44-3675-9d9c-ded47fe734ea",
    "path" : "\/Applications\/Ghostty.app\/Contents\/Frameworks\/Sparkle.framework\/Versions\/B\/Sparkle",
    "name" : "Sparkle",
    "CFBundleVersion" : "2039"
  },
  {
    "source" : "P",
    "arch" : "x86_64",
    "base" : 4484096000,
    "CFBundleShortVersionString" : "1.0.0",
    "CFBundleIdentifier" : "com.mitchellh.ghostty",
    "size" : 17342464,
    "uuid" : "f263bce6-c04b-3a3f-9a6b-297a11b22b9a",
    "path" : "\/Applications\/Ghostty.app\/Contents\/MacOS\/ghostty",
    "name" : "ghostty",
    "CFBundleVersion" : "8469"
  },
  {
    "size" : 0,
    "source" : "A",
    "base" : 0,
    "uuid" : "00000000-0000-0000-0000-000000000000"
  },
  {
    "source" : "P",
    "arch" : "x86_64",
    "base" : 140703413235712,
    "size" : 49144,
    "uuid" : "79ecab15-71f1-3d6a-8a96-0623e622205f",
    "path" : "\/usr\/lib\/system\/libsystem_pthread.dylib",
    "name" : "libsystem_pthread.dylib"
  },
  {
    "source" : "P",
    "arch" : "x86_64",
    "base" : 140703412994048,
    "size" : 241656,
    "uuid" : "b8a8e52d-b216-3bdb-a31b-1e34e9e9500b",
    "path" : "\/usr\/lib\/system\/libsystem_kernel.dylib",
    "name" : "libsystem_kernel.dylib"
  },
  {
    "source" : "P",
    "arch" : "x86_64",
    "base" : 140703411843072,
    "size" : 557048,
    "uuid" : "17b641ba-925c-39a9-aa43-ab3b0bcdfe01",
    "path" : "\/usr\/lib\/system\/libsystem_c.dylib",
    "name" : "libsystem_c.dylib"
  },
  {
    "source" : "P",
    "arch" : "x86_64",
    "base" : 140703411277824,
    "size" : 262144,
    "uuid" : "eff78fb7-b5a3-3efc-8c45-309184ae989f",
    "path" : "\/usr\/lib\/system\/libsystem_malloc.dylib",
    "name" : "libsystem_malloc.dylib"
  }
],
  "sharedCache" : {
  "base" : 140703408820224,
  "size" : 25769803776,
  "uuid" : "a7563e1d-7a04-3ed2-9587-044987205172"
},
  "vmSummary" : "ReadOnly portion of Libraries: Total=750.8M resident=0K(0%) swapped_out_or_unallocated=750.8M(100%)\nWritable regions: Total=576.0M written=0K(0%) resident=0K(0%) swapped_out=0K(0%) unallocated=576.0M(100%)\n\n                                VIRTUAL   REGION \nREGION TYPE                        SIZE    COUNT (non-coalesced) \n===========                     =======  ======= \nActivity Tracing                   256K        1 \nColorSync                          156K       22 \nKernel Alloc Once                    8K        1 \nMALLOC                           549.3M       23 \nMALLOC guard page                   32K        8 \nSTACK GUARD                       56.0M        6 \nStack                             26.0M        6 \nVM_ALLOCATE                        268K        4 \n__CTF                               824        1 \n__DATA                            28.0M      443 \n__DATA_CONST                      28.4M      273 \n__DATA_DIRTY                      1120K      137 \n__FONT_DATA                        2352        1 \n__LINKEDIT                       183.4M        4 \n__OBJC_RO                         71.9M        1 \n__OBJC_RW                         2201K        2 \n__TEXT                           567.4M      463 \nmapped file                       53.4M       10 \nshared memory                      784K       17 \n===========                     =======  ======= \nTOTAL                              1.5G     1423 \n",
  "legacyInfo" : {
  "threadTriggered" : {

  }
},
  "logWritingSignature" : "b25d039b738b502bda27386473daa1aae94daf51",
  "trialInfo" : {
  "rollouts" : [
    {
      "rolloutId" : "64c17a9925d75a7281053d4c",
      "factorPackIds" : {
        "SIRI_AUDIO_DISABLE_MEDIA_ENTITY_SYNC" : "64d29746ad29a465b3bbeace"
      },
      "deploymentId" : 240000002
    },
    {
      "rolloutId" : "645197bf528fbf3c3af54105",
      "factorPackIds" : {
        "SIRI_VALUE_INFERENCE_PERVASIVE_ENTITY_RESOLUTION" : "663e65b4a1526e1ca0e288a1"
      },
      "deploymentId" : 240000002
    }
  ],
  "experiments" : [
    {
      "treatmentId" : "45f4e2a5-551b-4bc2-a2dc-19c244dda8f8",
      "experimentId" : "6643969b3099cf28e049862f",
      "deploymentId" : 400000006
    }
  ]
}
}

Model: MacBookAir8,1, BootROM 2069.40.2.0.0 (iBridge: 22.16.11072.0.0,0), 2 processors, Dual-Core Intel Core i5, 1,6 GHz, 8 GB, SMC 
Graphics: Intel UHD Graphics 617, Intel UHD Graphics 617, Built-In
Display: Color LCD, 2560 x 1600 Retina, Main, MirrorOff, Online
Memory Module: BANK 0/DIMM0, 4 GB, LPDDR3, 2133 MHz, 0x80AD, 0x483943434E4E4E435054414C42522D4E5644
Memory Module: BANK 1/DIMM0, 4 GB, LPDDR3, 2133 MHz, 0x80AD, 0x483943434E4E4E435054414C42522D4E5644
AirPort: spairport_wireless_card_type_wifi (0x14E4, 0x843), wl0: Jul  8 2023 17:49:51 version 9.30.503.0.32.5.87 FWID 01-e74b9cb0
AirPort: 
Bluetooth: Version (null), 0 services, 0 devices, 0 incoming serial ports
Network Service: Wi-Fi, AirPort, en0
USB Device: USB30Bus
USB Device: T2Bus
USB Device: Apple Internal Keyboard / Trackpad
USB Device: Headset
USB Device: Ambient Light Sensor
USB Device: FaceTime HD Camera (Built-in)
USB Device: Apple T2 Controller
Thunderbolt Bus: MacBook Air, Apple Inc., 47.5
</pre>
</details>

**Comment by mitchellh** [comment, marked as answer]
I made an issue for this one #3202. Its not verified but your description is super clear enough that it warrants an issue.
